pkg/agent/agentclient/stream.go

```diff
@@ -168,49 +187,142 @@ func (c *RedialableAgentClient) Recv() (*agent.Packet, error) {
 	return pkt, nil
 }
 
-func (c *RedialableAgentClient) Connect() error {
-	if err := c.tryConnect(); err != nil {
-		return err
-	}
+// Connect makes the grpc dial to the proxy server. It returns the serverID
+// it connects to.
+func (c *RedialableAgentClient) Connect() (string, error) {
+	var err error
+	c.serverID, c.serverCount, c.conn, c.stream, err = c.tryConnect()
+	return c.serverID, err
+}
 
-	go c.probe()
-	return nil
+// The goal is to make the chance that client's Connect rpc call has never hit
+// the wanted server after "retries" times to be lower than 10^-2.
+func retryLimit(serverCount int) (retries int) {
+	switch serverCount {
+	case 1:
+		return 3 // to overcome transient errors
+	case 2:
+		return 3 + 7
+	case 3:
+		return 3 + 12
+	case 4:
+		return 3 + 17
+	case 5:
+		return 3 + 21
+	default:
+		// we don't expect HA server with more than 5 instances.
+		return 3 + 21
+	}
 }
 
 func (c *RedialableAgentClient) reconnect() {
-	klog.Info("start to connect...")
+	klog.Info("start to reconnect...")
 
 	var err error
-	var retry int
-
-	for retry < c.Retry {
-		if err = c.tryConnect(); err == nil {
-			klog.Info("connected")
+	var retry, limit int
+
+	limit = retryLimit(c.serverCount)
+	for retry < limit {
+		serverID, _, err, conn, stream := c.tryConnect()
+		switch {
+		case err != nil:
```

Comment:
Can we nest this in an if statement block? Some of the latter switch conditions are a bit difficult to read grouped within one statement.

```
if err != nil {
 ...
} else {
 ...
}
```